File: Main/150/286.cc
#include <vector>

using namespace std;

// idea: dfs from every gate
class Solution {
  public:
    void f(vector<vector<int>> &rooms, int r, int c, int d) {
        if (r < 0 || c < 0 ||
            r >= rooms.size() || c >= rooms[0].size() ||
            rooms[r][c] == 0 ||
            rooms[r][c] == -1)
            return;

        if (d < rooms[r][c]) {
            rooms[r][c] = d;
            f(rooms, r + 1, c, d + 1);
            f(rooms, r - 1, c, d + 1);
            f(rooms, r, c + 1, d + 1);
            f(rooms, r, c - 1, d + 1);
        }
    }
    void wallsAndGates(vector<vector<int>> &rooms) {
        for (int i = 0; i < rooms.size(); i++) {
            for (int j = 0; j < rooms[0].size(); j++) {
                if (rooms[i][j] == 0) {
                    f(rooms, i + 1, j, 1);
                    f(rooms, i - 1, j, 1);
                    f(rooms, i, j + 1, 1);
                    f(rooms, i, j - 1, 1);
                }
            }
        }
    }
};

File: src/150redo/Graphs/Go/286.go
package graphs

func wallsAndGates(rooms [][]int) {
	for i := 0; i < len(rooms); i++ {
		for j := 0; j < len(rooms[0]); j++ {
			if rooms[i][j] == 0 {
				dfs2(rooms, i+1, j, 1)
				dfs2(rooms, i-1, j, 1)
				dfs2(rooms, i, j+1, 1)
				dfs2(rooms, i, j-1, 1)
			}
		}
	}
}

func dfs2(rooms [][]int, r, c, d int) {
	if r < 0 || r >= len(rooms) || c < 0 || c >= len(rooms[0]) || rooms[r][c] == -1 || rooms[r][c] == 0 {
		return
	}

	if d < rooms[r][c] {
		rooms[r][c] = d
		dfs2(rooms, r+1, c, d+1)
		dfs2(rooms, r-1, c, d+1)
		dfs2(rooms, r, c+1, d+1)
		dfs2(rooms, r, c-1, d+1)
	}
}
